In [1]:
import pandas as pd
import numpy as np
import json
import matplotlib.pyplot as plt

In [235]:
#f = open('goodreads_books_mystery_thriller_crime.json')

In [236]:
#data = json.load(f)

In [237]:
df_full = pd.read_json('goodreads_books_mystery_thriller_crime.json')

df_full.head()

isbn  text_reviews_count     series country_code language_code  \
0  184737297X                  15   [169353]           US                 
1                              60  [1052227]           US           eng   
2                              23   [953679]           US           eng   
3  0854563903                   8   [408775]           US                 
4  8838920931                   3   [274410]           US           ita   

                                     popular_shelves        asin is_ebook  \
0  [{'count': '159', 'name': 'to-read'}, {'count'...                false   
1  [{'count': '54', 'name': 'currently-reading'},...  B01NCIKAQX     true   
2  [{'count': '90', 'name': 'to-read'}, {'count':...  B01ALOWJN0     true   
3  [{'count': '51', 'name': 'to-read'}, {'count':...                false   
4  [{'count': '48', 'name': 'to-read'}, {'count':...                false   

   average_rating kindle_asin  ... publication_month edition_information  \
0            3.93  B007YLTG5I  ...                 4                       
1            4.33  B01NCIKAQX  ...                                         
2            3.49  B01ALOWJN0  ...                                         
3            3.30              ...                12         Large Print   
4            3.54              ...                                         

  publication_year                                                url  \
0             2009  https://www.goodreads.com/book/show/6066814-cr...   
1                   https://www.goodreads.com/book/show/33394837-t...   
2                   https://www.goodreads.com/book/show/29074697-t...   
3             1975  https://www.goodreads.com/book/show/1902202.De...   
4             2006  https://www.goodreads.com/book/show/9671977-ar...   

                                           image_url   book_id ratings_count  \
0  https://images.gr-assets.com/books/1328724803m...   6066814           186   
1  https://images.gr-assets.com/books/1493114742m...  33394837           269   
2  https://s.gr-assets.com/assets/nophoto/book/11...  29074697           192   
3  https://s.gr-assets.com/assets/nophoto/book/11...   1902202            52   
4  https://images.gr-assets.com/books/1474788304m...   9671977            22   

    work_id                                              title  \
0   6243149          Crowner Royal (Crowner John Mystery, #13)   
1  54143148            The House of Memory (Pluto's Snitch #2)   
2  49305010  The Slaughtered Virgin of Zenopolis (Inspector...   
3   1903897            Dead in the Morning (Patrick Grant, #1)   
4   2152906                   Aristotele e i misteri di Eleusi   

                                title_without_series  
0          Crowner Royal (Crowner John Mystery, #13)  
1            The House of Memory (Pluto's Snitch #2)  
2  The Slaughtered Virgin of Zenopolis (Inspector...  
3            Dead in the Morning (Patrick Grant, #1)  
4                   Aristotele e i misteri di Eleusi  

[5 rows x 29 columns]

In [238]:
df = df_full

In [239]:
df.columns

Index(['isbn', 'text_reviews_count', 'series', 'country_code', 'language_code',
       'popular_shelves', 'asin', 'is_ebook', 'average_rating', 'kindle_asin',
       'similar_books', 'description', 'format', 'link', 'authors',
       'publisher', 'num_pages', 'publication_day', 'isbn13',
       'publication_month', 'edition_information', 'publication_year', 'url',
       'image_url', 'book_id', 'ratings_count', 'work_id', 'title',
       'title_without_series'],
      dtype='object')

In [240]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219235 entries, 0 to 219234
Data columns (total 29 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   isbn                  219235 non-null  object 
 1   text_reviews_count    219235 non-null  int64  
 2   series                219235 non-null  object 
 3   country_code          219235 non-null  object 
 4   language_code         219235 non-null  object 
 5   popular_shelves       219235 non-null  object 
 6   asin                  219235 non-null  object 
 7   is_ebook              219235 non-null  object 
 8   average_rating        219235 non-null  float64
 9   kindle_asin           219235 non-null  object 
 10  similar_books         219235 non-null  object 
 11  description           219235 non-null  object 
 12  format                219235 non-null  object 
 13  link                  219235 non-null  object 
 14  authors               219235 non-null  object 
 15  

Replacing blanks with null values for better analysis

In [241]:
df = df.replace(r'^\s*$', np.nan, regex=True)

In [242]:
df.isna().sum()

isbn                     84517
text_reviews_count           0
series                       0
country_code                 0
language_code            82451
popular_shelves              0
asin                    170799
is_ebook                     0
average_rating               0
kindle_asin              99160
similar_books                0
description              20747
format                   56845
link                         0
authors                      0
publisher                57932
num_pages                69828
publication_day          89035
isbn13                   67288
publication_month        77064
edition_information     197401
publication_year         54494
url                          0
image_url                    0
book_id                      0
ratings_count                0
work_id                      0
title                        0
title_without_series         0
dtype: int64

In [243]:
df.title.nunique(), df.book_id.nunique(), df.isbn.nunique()

(136843, 219235, 134718)

Note, the null+nunique for isbn comes to the total for book id. Therefore, we will use book id as our id information and drop the other id columns. Using book_id going forward as the id, will drop the others. 

In [244]:
#Potential drops
drop_otherid = ['isbn','isbn13', 'asin', 'kindle_asin']
drop_links = ['url','image_url', 'link']
drop_othertimes = ['publication_day']
drop_words = ['description']

In [245]:
df = df.drop(drop_otherid, axis=1)

#Drop links
df = df.drop(drop_links, axis=1)

#Drop Description
df = df.drop(drop_words, axis=1)

#Drop day published
df = df.drop(drop_othertimes, axis=1)

In [246]:
df.head()

text_reviews_count     series country_code language_code  \
0                  15   [169353]           US           NaN   
1                  60  [1052227]           US           eng   
2                  23   [953679]           US           eng   
3                   8   [408775]           US           NaN   
4                   3   [274410]           US           ita   

                                     popular_shelves is_ebook  average_rating  \
0  [{'count': '159', 'name': 'to-read'}, {'count'...    false            3.93   
1  [{'count': '54', 'name': 'currently-reading'},...     true            4.33   
2  [{'count': '90', 'name': 'to-read'}, {'count':...     true            3.49   
3  [{'count': '51', 'name': 'to-read'}, {'count':...    false            3.30   
4  [{'count': '48', 'name': 'to-read'}, {'count':...    false            3.54   

                                       similar_books     format  \
0  [439108, 522621, 116770, 1275927, 6202655, 840...  Hardcover   
1                                                 []        NaN   
2                                                 []        NaN   
3                                                 []  Hardcover   
4                                                 []  Paperback   

                                             authors                publisher  \
0               [{'author_id': '37778', 'role': ''}]      Simon & Schuster UK   
1              [{'author_id': '242185', 'role': ''}]                      NaN   
2  [{'author_id': '15104629', 'role': ''}, {'auth...  Amazon Digital Services   
3              [{'author_id': '190988', 'role': ''}]              Ulverscroft   
4  [{'author_id': '337108', 'role': ''}, {'author...                 Sellerio   

  num_pages publication_month edition_information publication_year   book_id  \
0       400                 4                 NaN             2009   6066814   
1       318               NaN                 NaN              NaN  33394837   
2       NaN               NaN                 NaN              NaN  29074697   
3       NaN                12         Large Print             1975   1902202   
4       659               NaN                 NaN             2006   9671977   

   ratings_count   work_id                                              title  \
0            186   6243149          Crowner Royal (Crowner John Mystery, #13)   
1            269  54143148            The House of Memory (Pluto's Snitch #2)   
2            192  49305010  The Slaughtered Virgin of Zenopolis (Inspector...   
3             52   1903897            Dead in the Morning (Patrick Grant, #1)   
4             22   2152906                   Aristotele e i misteri di Eleusi   

                                title_without_series  
0          Crowner Royal (Crowner John Mystery, #13)  
1            The House of Memory (Pluto's Snitch #2)  
2  The Slaughtered Virgin of Zenopolis (Inspector...  
3            Dead in the Morning (Patrick Grant, #1)  
4                   Aristotele e i misteri di Eleusi

## Let's clean up Series, Popular Shelves, and Authors

In [247]:
#Series first
df.series = df.series.str[0]

In [248]:
df.head()

text_reviews_count   series country_code language_code  \
0                  15   169353           US           NaN   
1                  60  1052227           US           eng   
2                  23   953679           US           eng   
3                   8   408775           US           NaN   
4                   3   274410           US           ita   

                                     popular_shelves is_ebook  average_rating  \
0  [{'count': '159', 'name': 'to-read'}, {'count'...    false            3.93   
1  [{'count': '54', 'name': 'currently-reading'},...     true            4.33   
2  [{'count': '90', 'name': 'to-read'}, {'count':...     true            3.49   
3  [{'count': '51', 'name': 'to-read'}, {'count':...    false            3.30   
4  [{'count': '48', 'name': 'to-read'}, {'count':...    false            3.54   

                                       similar_books     format  \
0  [439108, 522621, 116770, 1275927, 6202655, 840...  Hardcover   
1                                                 []        NaN   
2                                                 []        NaN   
3                                                 []  Hardcover   
4                                                 []  Paperback   

                                             authors                publisher  \
0               [{'author_id': '37778', 'role': ''}]      Simon & Schuster UK   
1              [{'author_id': '242185', 'role': ''}]                      NaN   
2  [{'author_id': '15104629', 'role': ''}, {'auth...  Amazon Digital Services   
3              [{'author_id': '190988', 'role': ''}]              Ulverscroft   
4  [{'author_id': '337108', 'role': ''}, {'author...                 Sellerio   

  num_pages publication_month edition_information publication_year   book_id  \
0       400                 4                 NaN             2009   6066814   
1       318               NaN                 NaN              NaN  33394837   
2       NaN               NaN                 NaN              NaN  29074697   
3       NaN                12         Large Print             1975   1902202   
4       659               NaN                 NaN             2006   9671977   

   ratings_count   work_id                                              title  \
0            186   6243149          Crowner Royal (Crowner John Mystery, #13)   
1            269  54143148            The House of Memory (Pluto's Snitch #2)   
2            192  49305010  The Slaughtered Virgin of Zenopolis (Inspector...   
3             52   1903897            Dead in the Morning (Patrick Grant, #1)   
4             22   2152906                   Aristotele e i misteri di Eleusi   

                                title_without_series  
0          Crowner Royal (Crowner John Mystery, #13)  
1            The House of Memory (Pluto's Snitch #2)  
2  The Slaughtered Virgin of Zenopolis (Inspector...  
3            Dead in the Morning (Patrick Grant, #1)  
4                   Aristotele e i misteri di Eleusi

In [249]:
#Now Authors
df.authors.head()

0                 [{'author_id': '37778', 'role': ''}]
1                [{'author_id': '242185', 'role': ''}]
2    [{'author_id': '15104629', 'role': ''}, {'auth...
3                [{'author_id': '190988', 'role': ''}]
4    [{'author_id': '337108', 'role': ''}, {'author...
Name: authors, dtype: object

In [250]:
auths = df.authors.apply(pd.Series)
auths.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219235 entries, 0 to 219234
Data columns (total 51 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   0       219235 non-null  object
 1   1       44710 non-null   object
 2   2       7065 non-null    object
 3   3       2113 non-null    object
 4   4       1276 non-null    object
 5   5       1022 non-null    object
 6   6       908 non-null     object
 7   7       820 non-null     object
 8   8       758 non-null     object
 9   9       675 non-null     object
 10  10      601 non-null     object
 11  11      553 non-null     object
 12  12      506 non-null     object
 13  13      457 non-null     object
 14  14      404 non-null     object
 15  15      332 non-null     object
 16  16      295 non-null     object
 17  17      258 non-null     object
 18  18      233 non-null     object
 19  19      209 non-null     object
 20  20      172 non-null     object
 21  21      141 non-null     object
 

In [251]:
#For space, we will only considered 3 author listings
au1 = auths[0].apply(pd.Series)
au2 = auths[1].apply(pd.Series)
au3 = auths[2].apply(pd.Series)

In [252]:
au1.head()

author_id role
0     37778     
1    242185     
2  15104629     
3    190988     
4    337108

In [253]:
df['1st_au_id'] = au1.author_id
df['1st_au_role'] = au1.role
df['2nd_au_id'] = au2.author_id
df['2nd_au_role'] = au2.role
df['3rd_au_id'] = au3.author_id
df['3rd_au_role'] = au3.role

In [254]:
df.head()

text_reviews_count   series country_code language_code  \
0                  15   169353           US           NaN   
1                  60  1052227           US           eng   
2                  23   953679           US           eng   
3                   8   408775           US           NaN   
4                   3   274410           US           ita   

                                     popular_shelves is_ebook  average_rating  \
0  [{'count': '159', 'name': 'to-read'}, {'count'...    false            3.93   
1  [{'count': '54', 'name': 'currently-reading'},...     true            4.33   
2  [{'count': '90', 'name': 'to-read'}, {'count':...     true            3.49   
3  [{'count': '51', 'name': 'to-read'}, {'count':...    false            3.30   
4  [{'count': '48', 'name': 'to-read'}, {'count':...    false            3.54   

                                       similar_books     format  \
0  [439108, 522621, 116770, 1275927, 6202655, 840...  Hardcover   
1                                                 []        NaN   
2                                                 []        NaN   
3                                                 []  Hardcover   
4                                                 []  Paperback   

                                             authors  ... ratings_count  \
0               [{'author_id': '37778', 'role': ''}]  ...           186   
1              [{'author_id': '242185', 'role': ''}]  ...           269   
2  [{'author_id': '15104629', 'role': ''}, {'auth...  ...           192   
3              [{'author_id': '190988', 'role': ''}]  ...            52   
4  [{'author_id': '337108', 'role': ''}, {'author...  ...            22   

    work_id                                              title  \
0   6243149          Crowner Royal (Crowner John Mystery, #13)   
1  54143148            The House of Memory (Pluto's Snitch #2)   
2  49305010  The Slaughtered Virgin of Zenopolis (Inspector...   
3   1903897            Dead in the Morning (Patrick Grant, #1)   
4   2152906                   Aristotele e i misteri di Eleusi   

                                title_without_series 1st_au_id  1st_au_role  \
0          Crowner Royal (Crowner John Mystery, #13)     37778                
1            The House of Memory (Pluto's Snitch #2)    242185                
2  The Slaughtered Virgin of Zenopolis (Inspector...  15104629                
3            Dead in the Morning (Patrick Grant, #1)    190988                
4                   Aristotele e i misteri di Eleusi    337108                

   2nd_au_id  2nd_au_role 3rd_au_id 3rd_au_role  
0        NaN          NaN       NaN         NaN  
1        NaN          NaN       NaN         NaN  
2   14425713                    NaN         NaN  
3        NaN          NaN       NaN         NaN  
4    4585413   Translator   1093599       Notes  

[5 rows x 26 columns]

In [255]:
#Dropping authors column
df.drop(columns=['authors'], inplace=True)

In [256]:
#df=df_full

----------Got Popular Shelves separated. Will address if time.-----

In [257]:
df['popular_shelves']

0         [{'count': '159', 'name': 'to-read'}, {'count'...
1         [{'count': '54', 'name': 'currently-reading'},...
2         [{'count': '90', 'name': 'to-read'}, {'count':...
3         [{'count': '51', 'name': 'to-read'}, {'count':...
4         [{'count': '48', 'name': 'to-read'}, {'count':...
                                ...                        
219230    [{'count': '2422', 'name': 'to-read'}, {'count...
219231    [{'count': '497', 'name': 'to-read'}, {'count'...
219232    [{'count': '735', 'name': 'to-read'}, {'count'...
219233    [{'count': '36', 'name': 'to-read'}, {'count':...
219234    [{'count': '702', 'name': 'to-read'}, {'count'...
Name: popular_shelves, Length: 219235, dtype: object

In [258]:
#Now onto Popular Shelves

#ps = df.popular_shelves.apply(pd.Series)

In [259]:
#ps

In [260]:
#ps_t = ps[:10]
#ps_t

#For future use if I have time
ps_df = pd.DataFrame()
for i in ps_t.columns:
    psi = ps_t[i].apply(pd.Series)
    cols = ['ps_count_'+str(i),'ps_name_'+str(i)]
    psi.columns = cols
    ps_df = pd.concat([ps_df,psi], axis=1)
ps_df

Similarly, I will skip 'similar_books' for now, in the interest of time. 

---------Popular shelves ends here------

In [261]:
df.columns

Index(['text_reviews_count', 'series', 'country_code', 'language_code',
       'popular_shelves', 'is_ebook', 'average_rating', 'similar_books',
       'format', 'publisher', 'num_pages', 'publication_month',
       'edition_information', 'publication_year', 'book_id', 'ratings_count',
       'work_id', 'title', 'title_without_series', '1st_au_id', '1st_au_role',
       '2nd_au_id', '2nd_au_role', '3rd_au_id', '3rd_au_role'],
      dtype='object')

In [262]:
df_ps = df['popular_shelves']
df_sb = df['similar_books']

In [263]:
df.drop(columns=['popular_shelves','similar_books'],inplace=True)

In [264]:
df.head()

text_reviews_count   series country_code language_code is_ebook  \
0                  15   169353           US           NaN    false   
1                  60  1052227           US           eng     true   
2                  23   953679           US           eng     true   
3                   8   408775           US           NaN    false   
4                   3   274410           US           ita    false   

   average_rating     format                publisher num_pages  \
0            3.93  Hardcover      Simon & Schuster UK       400   
1            4.33        NaN                      NaN       318   
2            3.49        NaN  Amazon Digital Services       NaN   
3            3.30  Hardcover              Ulverscroft       NaN   
4            3.54  Paperback                 Sellerio       659   

  publication_month  ... ratings_count   work_id  \
0                 4  ...           186   6243149   
1               NaN  ...           269  54143148   
2               NaN  ...           192  49305010   
3                12  ...            52   1903897   
4               NaN  ...            22   2152906   

                                               title  \
0          Crowner Royal (Crowner John Mystery, #13)   
1            The House of Memory (Pluto's Snitch #2)   
2  The Slaughtered Virgin of Zenopolis (Inspector...   
3            Dead in the Morning (Patrick Grant, #1)   
4                   Aristotele e i misteri di Eleusi   

                                title_without_series  1st_au_id 1st_au_role  \
0          Crowner Royal (Crowner John Mystery, #13)      37778               
1            The House of Memory (Pluto's Snitch #2)     242185               
2  The Slaughtered Virgin of Zenopolis (Inspector...   15104629               
3            Dead in the Morning (Patrick Grant, #1)     190988               
4                   Aristotele e i misteri di Eleusi     337108               

  2nd_au_id 2nd_au_role 3rd_au_id 3rd_au_role  
0       NaN         NaN       NaN         NaN  
1       NaN         NaN       NaN         NaN  
2  14425713                   NaN         NaN  
3       NaN         NaN       NaN         NaN  
4   4585413  Translator   1093599       Notes  

[5 rows x 23 columns]

DF is generally cleaned up with book_id being selected as main reference due to n/a values, other ids dropped, day dropped, and due to size constraints, popular_shelves and similar_books also dropped. 3 other columns are kept. Now onto data cleaning. 

In [265]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219235 entries, 0 to 219234
Data columns (total 23 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   text_reviews_count    219235 non-null  int64  
 1   series                131828 non-null  object 
 2   country_code          219235 non-null  object 
 3   language_code         136784 non-null  object 
 4   is_ebook              219235 non-null  object 
 5   average_rating        219235 non-null  float64
 6   format                162390 non-null  object 
 7   publisher             161303 non-null  object 
 8   num_pages             149407 non-null  object 
 9   publication_month     142171 non-null  object 
 10  edition_information   21834 non-null   object 
 11  publication_year      164741 non-null  object 
 12  book_id               219235 non-null  int64  
 13  ratings_count         219235 non-null  int64  
 14  work_id               219235 non-null  int64  
 15  

Looking at language and country_code first as we want to focus on english only. 

In [266]:
df.language_code.nunique()

86

In [267]:
df.language_code.unique()

array([nan, 'eng', 'ita', 'en-GB', 'en-CA', 'gre', 'lit', 'fin', 'en-US',
       'per', 'bul', 'nl', 'spa', 'ger', 'swe', 'pes', 'cze', 'en', 'por',
       'fre', 'nor', 'zho', 'ind', 'rum', 'slo', 'lav', 'hun', 'afr',
       'heb', 'dan', 'pol', 'srp', 'tur', 'kat', 'jpn', 'ben', 'tha',
       'nob', 'vie', 'rus', 'ara', 'cat', 'es-MX', 'ukr', 'kan', 'scr',
       'msa', 'est', 'isl', 'pra', 'urd', 'mal', 'tam', 'glg', 'hin',
       'vls', 'din', 'pt-BR', 'slv', 'aze', 'nld', 'mkd', 'aus', 'mul',
       'eus', 'guj', 'mon', 'hye', 'fil', 'tgl', 'mus', 'abk', 'nno',
       'kor', 'frs', '--', 'mlt', 'grc', 'egy', 'jbo', 'mar', 'sin',
       'peo', 'sqi', 'bos', 'tel', 'ady'], dtype=object)

In [268]:
english = ['eng', 'en-GB', 'en-CA','en-US','en',np.nan, '--']

In [269]:
df = df.loc[df['language_code'].isin(english)]

In [270]:
df.country_code.unique()

array(['US'], dtype=object)

In [271]:
#Only US still there. 
df.country_code.isnull().sum()

0

In [272]:
df.language_code.isnull().sum()

82451

In [273]:
#Let's look at '--' in case it's not english
df[df.language_code=='--']

text_reviews_count  series country_code language_code is_ebook  \
85197                   13     NaN           US            --    false   
192814                   1  343793           US            --    false   
210543                   3     NaN           US            --    false   

        average_rating                 format             publisher num_pages  \
85197             3.91              Paperback             Alfaguara       544   
192814            3.75  Mass Market Paperback  Hodder and Stoughton       256   
210543            3.90              Paperback           Nova knjiga       268   

       publication_month  ... ratings_count   work_id  \
85197                 11  ...           143   2302976   
192814               NaN  ...             1   2275111   
210543               NaN  ...            13  40783433   

                          title     title_without_series  1st_au_id  \
85197          La reina del sur         La reina del sur      40398   
192814  The Saint Bids Diamonds  The Saint Bids Diamonds      36260   
210543             Central park             Central park     282080   

       1st_au_role 2nd_au_id 2nd_au_role 3rd_au_id 3rd_au_role  
85197                    NaN         NaN       NaN         NaN  
192814                   NaN         NaN       NaN         NaN  
210543               8585350  Translator       NaN         NaN  

[3 rows x 23 columns]

In [274]:
dr_sp = df[(df.work_id == 2302976) & (df.title == 'La reina del sur') & (df.language_code != 'eng')].index

In [275]:
df.drop(dr_sp, inplace=True)

In [276]:
df.language_code[df.language_code=='--'] = np.nan

<ipython-input-276-cf63f822c440>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.language_code[df.language_code=='--'] = np.nan


Now looking at duplicate values for future reference. Notes to decide whether to join them or not. Many duplicates are from different publishers and possibly translators. Also audio vs. ebook vs. traditional books. Just getting the duplicate list now so we have it for reference. 

In [277]:
df.title.value_counts()

The Adventures of Sherlock Holmes                                                93
Rebecca                                                                          81
The Hound of the Baskervilles                                                    71
The Woman in White                                                               68
And Then There Were None                                                         67
                                                                                 ..
SAVANNAH GONE: A Ray Fontaine Mystery                                             1
The Galleons' Grave: Henry Gresham and the Spanish Armada (Henry Gresham, #3)     1
Nighthawk: African Ice                                                            1
Spooky House (A Billie B Mystery #1)                                              1
Gem of a Ghost                                                                    1
Name: title, Length: 107563, dtype: int64

In [278]:
df.work_id.value_counts()

3311984     115
1222101     112
1997473      88
3038872      85
1303710      82
           ... 
10487496      1
818887        1
52934334      1
15963834      1
56360960      1
Name: work_id, Length: 83771, dtype: int64

By titles, there's 107563 unique books, but by work_id, there's only 83,771. Many of these likely have differences. First, we will sort out the other characteristics: format, edition information.

In [279]:
df.format.value_counts().head(20)

Paperback                58201
Hardcover                23801
ebook                    14977
Kindle Edition           12819
Audio CD                  6666
Mass Market Paperback     5072
Audiobook                 3048
Audible Audio             2024
Audio                     1823
Unknown Binding            518
Audio Cassette             232
MP3 CD                     193
Library Binding             92
Trade Paperback             91
Nook                        56
paperback                   54
Leather Bound               25
Hardback                    19
hardcover                   13
Digital Audio               10
Name: format, dtype: int64

In [280]:
#Making all formats lower case
df.format = df.format.str.lower()

In [281]:
df.edition_information[df.format.str.contains('large', regex=True, na=False)]

4109      Large Print
12612             NaN
13089     Large Print
14336             NaN
18682     Large Print
43914     Large Print
54156     Large Print
58129             NaN
68475             NaN
71850             NaN
83647             NaN
101375            NaN
113964            NaN
127420    Large Print
138584    Large Print
148263            NaN
166227            NaN
168611            NaN
176179            NaN
201153    Large Print
201205            NaN
202240            NaN
210281    Large Print
212841            NaN
213989            NaN
215329            NaN
Name: edition_information, dtype: object

In [282]:
#Setting NaN large print edtion to Large Print 
df.edition_information[df.format.str.contains('large', regex=True, na=False)] = 'Large Print'

<ipython-input-282-1742681b4d41>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.edition_information[df.format.str.contains('large', regex=True, na=False)] = 'Large Print'


In [283]:
df.edition_information[df.format.str.contains('large', regex=True, na=False)]

4109      Large Print
12612     Large Print
13089     Large Print
14336     Large Print
18682     Large Print
43914     Large Print
54156     Large Print
58129     Large Print
68475     Large Print
71850     Large Print
83647     Large Print
101375    Large Print
113964    Large Print
127420    Large Print
138584    Large Print
148263    Large Print
166227    Large Print
168611    Large Print
176179    Large Print
201153    Large Print
201205    Large Print
202240    Large Print
210281    Large Print
212841    Large Print
213989    Large Print
215329    Large Print
Name: edition_information, dtype: object

In [284]:
df.format[df.format.str.contains('odc|odio', regex=True, na=False)].value_counts()

podiobook              5
podcast                2
podcast / audiobook    1
podio book             1
podcast/podiobooks     1
Name: format, dtype: int64

In [285]:
df.format[df.format.str.contains('odc|odio', regex=True, na=False)] = 'Podcast'

<ipython-input-285-8d5d17f14051>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.format[df.format.str.contains('odc|odio', regex=True, na=False)] = 'Podcast'


In [286]:
df.format[df.format.str.contains('and|;|/|mul|,', regex=True, na=False)].value_counts()

paperback and ebook                      4
paperback, ebook                         4
kindle edition and paperback             4
ebook and paperback                      3
paperback and kindle                     3
ebook &amp; paperback                    2
ebook, paperback                         2
electronic &amp; print                   1
hardcover, book club edition             1
paperback and e-book                     1
e-book &amp; paperback                   1
ebook and trade paperback                1
paperback, kindle                        1
paperback/kindle                         1
audible, usa download                    1
ebook, print                             1
paperback, ebook, audible                1
kindle, paperback                        1
paperback, e-book                        1
true first ed hc, re-bound by library    1
ebook, paperback, audible audio          1
print and download                       1
print and ebook                          1
ebook and p

In [287]:
def find_formats(dataframe,col,strings):
    res = dataframe[col].loc[dataframe[col].str.contains(strings,regex=True,na=False)].value_counts()
    print(strings,'results are:\n',res)
    

In [288]:
def change_formats(dataframe,col,strings,new_name):
    dataframe[col].loc[dataframe[col].str.contains(strings,regex=True,na=False)] = new_name
    res = dataframe[col].loc[dataframe[col].str.contains(strings,regex=True,na=False)].value_counts()
    print(res)

In [289]:
find_formats(df, "format", 'and|;|/|mul|,')

and|;|/|mul|, results are:
 paperback and ebook                      4
paperback, ebook                         4
kindle edition and paperback             4
ebook and paperback                      3
paperback and kindle                     3
ebook &amp; paperback                    2
ebook, paperback                         2
electronic &amp; print                   1
hardcover, book club edition             1
paperback and e-book                     1
e-book &amp; paperback                   1
ebook and trade paperback                1
paperback, kindle                        1
paperback/kindle                         1
audible, usa download                    1
ebook, print                             1
paperback, ebook, audible                1
kindle, paperback                        1
paperback, e-book                        1
true first ed hc, re-bound by library    1
ebook, paperback, audible audio          1
print and download                       1
print and ebook           

In [290]:
find_formats(df, "format", 'hardcover, book club edition|hc')

hardcover, book club edition|hc results are:
 true first ed hc, re-bound by library    1
hardcover, book club edition             1
Name: format, dtype: int64


In [291]:
change_formats(df, "format", 'hardcover, book club edition|hc', "Hardcover")

Series([], Name: format, dtype: int64)


/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [292]:
find_formats(df, "format", 'audible, usa')

audible, usa results are:
 audible, usa download    1
Name: format, dtype: int64


In [293]:
change_formats(df, "format", 'audible, usa', "Audible")

Series([], Name: format, dtype: int64)


In [294]:
find_formats(df, "format", 'and|;|/|mul|,')

and|;|/|mul|, results are:
 paperback and ebook                4
kindle edition and paperback       4
paperback, ebook                   4
paperback and kindle               3
ebook and paperback                3
ebook &amp; paperback              2
ebook, paperback                   2
ebook, print                       1
paperback, kindle                  1
paperback, e-book                  1
electronic &amp; print             1
ebook, paperback, audible audio    1
ebook and trade paperback          1
e-book &amp; paperback             1
paperback/kindle                   1
paperback, ebook, audible          1
kindle, paperback                  1
paperback and e-book               1
print and ebook                    1
print and download                 1
ebook and print                    1
ebook&amp;paperback                1
paperback &amp; pdf                1
print/kindle                       1
multiple formats                   1
crime and investigation            1
paperback 

In [295]:
change_formats(df, "format", 'and|;|/|mul|,', "Multiple")

Series([], Name: format, dtype: int64)


In [296]:
find_formats(df, "format", 'hard')

hard results are:
 hardcover                    23815
hardback                        27
hardcover - large print          3
hard cover                       3
hardcover (large print)          2
hardcover-lp                     2
faux leather hardcover           1
large print hardcover            1
limited edition hardcover        1
original uk hardback             1
hardover                         1
Name: format, dtype: int64


In [297]:
change_formats(df, "format", 'hard', "Hardcover")

Series([], Name: format, dtype: int64)


In [298]:
find_formats(df, "format", 'oft|ape')

oft|ape results are:
 paperback                          58255
mass market paperback               5079
trade paperback                       96
softcover                              7
paper                                  6
soft cover                             4
paperback manga                        4
paperback - large print                3
tradepaperback                         2
uk paperback                           2
perfect paperback                      2
mass market paperback - reissue        1
paperback - c format                   1
(paperback)                            1
export trade paperback                 1
trade paper back                       1
vhs tape                               1
large print paperback                  1
pocket paperback                       1
mass-market paperback                  1
trade paperbck                         1
boxed set (paperback)                  1
open market paperback                  1
bookpaper                          

In [299]:
change_formats(df, "format", 'oft|ape', "Paperback")

Paperback    63473
Name: format, dtype: int64


In [300]:
find_formats(df, "format", 'kind')

kind results are:
 kindle edition            12819
kindle                        3
kindel edition                1
ebook (version kindle)        1
kindle ebook                  1
Name: format, dtype: int64


In [301]:
change_formats(df, "format", 'kind', "Kindle")

Series([], Name: format, dtype: int64)


In [302]:
find_formats(df, "format", 'audib')

audib results are:
 audible audio                   2025
audiobook - audible download       1
audible mp3 download               1
mp3 audible                        1
Name: format, dtype: int64


In [303]:
change_formats(df, "format", 'audib', 'Audible')

Series([], Name: format, dtype: int64)


In [304]:
find_formats(df, "format", 'cass|diskette|vhs')

cass|diskette|vhs results are:
 audio cassette                     233
audiocassette                        2
diskette                             1
audiocassette (unabridged)           1
audio cassette (libr.) (unabr.)      1
audio cassette - abridged            1
audio cassette - unabridged          1
Name: format, dtype: int64


In [305]:
change_formats(df, "format", 'cass', "Cassette")

Series([], Name: format, dtype: int64)


In [306]:
find_formats(df, "format", 'aud|mp3|cd')

aud|mp3|cd results are:
 audio cd                          6669
audiobook                         3050
audio                             1828
mp3 cd                             193
digital audio                       10
mp3 book                             5
eaudiobook                           3
audio book                           2
cd-rom                               2
digital audio (unabridged)           2
mp3                                  2
audio cd (unabr.)                    1
cd                                   1
abriged audio cd                     1
mp3 audio                            1
mp3 audiobook                        1
audiobook unabridged                 1
mp3 download                         1
playaway audio                       1
unabriged audio cd                   1
mp3 audio download                   1
audio playaway                       1
preloaded digital audio player       1
Name: format, dtype: int64


In [307]:
change_formats(df, "format", 'aud|mp3|cd', "Audiobook")

Series([], Name: format, dtype: int64)


In [308]:
df.format.value_counts()[df.format.value_counts()>20]

Paperback          63473
Hardcover          23859
ebook              14979
Kindle             12825
Audiobook          11778
Audible             2029
unknown binding      518
Cassette             239
library binding       92
nook                  56
Multiple              47
leather bound         25
Name: format, dtype: int64

In [309]:
find_formats(df, "format", 'noo|eb')

noo|eb results are:
 ebook               14979
nook                   56
web serial novel        1
online freebie          1
webtoon                 1
Name: format, dtype: int64


In [310]:
change_formats(df, "format", 'noo|eb', "Ebook")

/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


Series([], Name: format, dtype: int64)


In [311]:
df.format.value_counts()[df.format.value_counts()>3]

Paperback            63473
Hardcover            23859
Ebook                15038
Kindle               12825
Audiobook            11778
Audible               2029
unknown binding        518
Cassette               239
library binding         92
Multiple                47
leather bound           25
large print             16
Podcast                 10
capa mole                9
unbound                  8
leatherette bound        7
board book               4
cloth                    4
Name: format, dtype: int64

In [312]:
#The rest will be under Unknown_other
format_mainlist = ['Paperback','Hardcover','Ebook','Kindle','Audiobook','Audible', 'Podcast','Cassette']

In [318]:
df.format[(df.format.isin(format_mainlist) == False) & (df.is_ebook == 'true')] = 'Ebook'

<ipython-input-318-dcff747c0eb8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.format[(df.format.isin(format_mainlist) == False) & (df.is_ebook == 'true')] = 'Ebook'


In [319]:
df.format[df.format.isin(format_mainlist) == False] = 'Other_Unknown'

<ipython-input-319-ecd15a3639e9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.format[df.format.isin(format_mainlist) == False] = 'Other_Unknown'


In [320]:
df.format.value_counts()

Paperback        63473
Ebook            49892
Hardcover        23859
Other_Unknown    18797
Kindle           12825
Audiobook        11778
Audible           2029
Cassette           239
Podcast             10
Name: format, dtype: int64

In [321]:
#Drop Cassette/outdated technology
df.drop(df[df.format == "Cassette"].index, inplace=True)

In [322]:
#df.to_pickle('./Books_wrangled.pkl')

In [535]:
#df = pd.read_pickle('./Books_wrangled.pkl')

In [536]:
df.head()

text_reviews_count   series country_code language_code is_ebook  \
0                  15   169353           US           NaN    false   
1                  60  1052227           US           eng     true   
2                  23   953679           US           eng     true   
3                   8   408775           US           NaN    false   
5                   3   199039           US           eng    false   

   average_rating     format                publisher num_pages  \
0            3.93  Hardcover      Simon & Schuster UK       400   
1            4.33      Ebook                      NaN       318   
2            3.49      Ebook  Amazon Digital Services       NaN   
3            3.30  Hardcover              Ulverscroft       NaN   
5            3.96  Audiobook              HarperAudio       NaN   

  publication_month  ... ratings_count   work_id  \
0                 4  ...           186   6243149   
1               NaN  ...           269  54143148   
2               NaN  ...           192  49305010   
3                12  ...            52   1903897   
5                 3  ...             5   2288775   

                                               title  \
0          Crowner Royal (Crowner John Mystery, #13)   
1            The House of Memory (Pluto's Snitch #2)   
2  The Slaughtered Virgin of Zenopolis (Inspector...   
3            Dead in the Morning (Patrick Grant, #1)   
5                              A Murder is Announced   

                                title_without_series  1st_au_id 1st_au_role  \
0          Crowner Royal (Crowner John Mystery, #13)      37778               
1            The House of Memory (Pluto's Snitch #2)     242185               
2  The Slaughtered Virgin of Zenopolis (Inspector...   15104629               
3            Dead in the Morning (Patrick Grant, #1)     190988               
5                              A Murder is Announced     123715               

  2nd_au_id 2nd_au_role 3rd_au_id 3rd_au_role  
0       NaN         NaN       NaN         NaN  
1       NaN         NaN       NaN         NaN  
2  14425713                   NaN         NaN  
3       NaN         NaN       NaN         NaN  
5    552417    Narrator       NaN         NaN  

[5 rows x 23 columns]

Now to clean up Author and Role data, similar to format

In [537]:
df['1st_au_role'] = df['1st_au_role'].str.lower()
df['2nd_au_role'] = df['2nd_au_role'].str.lower()
df['3rd_au_role'] = df['3rd_au_role'].str.lower()

In [538]:
df['1st_au_role'].value_counts()

                            180478
editor                        1003
pseudonym                      461
author                          67
creator                         55
                             ...  
simon vance                      1
written as mary roberts          1
original screenplay              1
text, illustrator                1
narrated by: lorna raver         1
Name: 1st_au_role, Length: 142, dtype: int64

In [539]:
find_formats(df,'3rd_au_role',"uth")

uth results are:
 attributed author    1
utho                 1
original author      1
co-author            1
Name: 3rd_au_role, dtype: int64


In [540]:
change_formats(df,'2nd_au_role','afterword / author interview',"Support")

Series([], Name: 2nd_au_role, dtype: int64)


/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [541]:
find_formats(df,'2nd_au_role',"uth")

uth results are:
 co-author             12
author's true name     9
editor/author          5
joint author           2
original author        2
house author           1
editor, author         1
author/illustrator     1
original_author        1
co- author             1
ghost author           1
author, artist         1
Name: 2nd_au_role, dtype: int64


In [542]:
change_formats(df,'1st_au_role',"uth","Author")
change_formats(df,'2nd_au_role',"uth","Author")
change_formats(df,'3rd_au_role',"uth","Author")

Author    131
Name: 1st_au_role, dtype: int64
Author    37
Name: 2nd_au_role, dtype: int64
Author    4
Name: 3rd_au_role, dtype: int64


In [543]:
find_formats(df,'3rd_au_role',"ito")

ito results are:
 editor                  153
series editor            14
editor/contributor        3
founding editor           3
editor, translator        2
editor / contributor      1
editor                    1
 editor                   1
Name: 3rd_au_role, dtype: int64


In [544]:
change_formats(df,'1st_au_role',"ito","Editor")
change_formats(df,'2nd_au_role',"ito","Editor")
change_formats(df,'3rd_au_role',"ito","Editor")

Editor    1079
Name: 1st_au_role, dtype: int64
Editor    908
Name: 2nd_au_role, dtype: int64
Editor    178
Name: 3rd_au_role, dtype: int64


In [545]:
find_formats(df,'1st_au_role',"ontr")

ontr results are:
 contributor                  49
contributor & redactional     2
contributor and forward       1
translator & contributor      1
Name: 1st_au_role, dtype: int64


In [546]:
change_formats(df,'1st_au_role',"ontr","Contributor")
change_formats(df,'2nd_au_role',"ontr","Contributor")
change_formats(df,'3rd_au_role',"ontr","Contributor")

Contributor    53
Name: 1st_au_role, dtype: int64
Contributor    317
Name: 2nd_au_role, dtype: int64
Contributor    291
Name: 3rd_au_role, dtype: int64


In [547]:
find_formats(df,'2nd_au_role',"host")

host results are:
 ghostwriter      34
ghost writer      7
ghost writer      1
Name: 2nd_au_role, dtype: int64


In [548]:
change_formats(df,'1st_au_role',"host","GhostWriter")
change_formats(df,'2nd_au_role',"host","GhostWriter")
change_formats(df,'3rd_au_role',"host","GhostWriter")

Series([], Name: 1st_au_role, dtype: int64)
GhostWriter    42
Name: 2nd_au_role, dtype: int64
Series([], Name: 3rd_au_role, dtype: int64)


In [549]:
find_formats(df,'3rd_au_role',"eat")

eat results are:
 series creator    71
creator           29
created by         1
Name: 3rd_au_role, dtype: int64


In [550]:
change_formats(df,'1st_au_role',"eat","Creator")
change_formats(df,'2nd_au_role',"eat","Creator")
change_formats(df,'3rd_au_role',"eat","Creator")

Creator    68
Name: 1st_au_role, dtype: int64
Creator    226
Name: 2nd_au_role, dtype: int64
Creator    101
Name: 3rd_au_role, dtype: int64


In [551]:
find_formats(df,'2nd_au_role',"eu")

eu results are:
 pseudonym     49
pseudonym     12
pseud.        11
Name: 2nd_au_role, dtype: int64


In [552]:
change_formats(df,'1st_au_role',"eu","Pseudonym")
change_formats(df,'2nd_au_role',"eu","Pseudonym")
change_formats(df,'3rd_au_role',"eu","Pseudonym")

Pseudonym    468
Name: 1st_au_role, dtype: int64
Pseudonym    72
Name: 2nd_au_role, dtype: int64
Pseudonym    7
Name: 3rd_au_role, dtype: int64


In [553]:
find_formats(df,'3rd_au_role',"arr")

arr results are:
 narrator           934
narrator            16
narrated by          8
as harriet vane      1
narratot             1
Name: 3rd_au_role, dtype: int64


In [554]:
change_formats(df,'1st_au_role',"arr","Narrator")
change_formats(df,'2nd_au_role',"arr","Narrator")
change_formats(df,'3rd_au_role',"arr","Narrator")

Narrator    39
Name: 1st_au_role, dtype: int64
Narrator    7427
Name: 2nd_au_role, dtype: int64
Narrator    960
Name: 3rd_au_role, dtype: int64


In [555]:
df['1st_au_role'].value_counts()[df['1st_au_role'].value_counts()>20]

                        180478
Editor                    1079
Pseudonym                  468
Author                     131
Creator                     68
Contributor                 53
house name                  41
Narrator                    39
adapter                     34
adapter/novelization        30
Name: 1st_au_role, dtype: int64

In [556]:
find_formats(df,'3rd_au_role',"ans")

ans results are:
 translator     294
translation     63
Name: 3rd_au_role, dtype: int64


In [557]:
change_formats(df,'1st_au_role',"ans|ubersetzer|traduc","Translator")
change_formats(df,'2nd_au_role',"ans|ubersetzer|traduc","Translator")
change_formats(df,'3rd_au_role',"ans|ubersetzer|traduc","Translator")

Translator    8
Name: 1st_au_role, dtype: int64


/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


Translator    3074
Name: 2nd_au_role, dtype: int64
Translator    370
Name: 3rd_au_role, dtype: int64


In [558]:
change_formats(df,'1st_au_role',"apt","Adapter")
change_formats(df,'2nd_au_role',"apt","Adapter")
change_formats(df,'3rd_au_role',"apt","Adapter")

Adapter    106
Name: 1st_au_role, dtype: int64
Adapter    16
Name: 2nd_au_role, dtype: int64
Adapter    7
Name: 3rd_au_role, dtype: int64


In [559]:
df['1st_au_role'].value_counts()[df['1st_au_role'].value_counts()>10]

                180478
Editor            1079
Pseudonym          468
Author             131
Adapter            106
Creator             68
Contributor         53
house name          41
Narrator            39
reading             20
pen name            18
preface             14
writer              12
introduction        12
Name: 1st_au_role, dtype: int64

In [560]:
find_formats(df,'3rd_au_role',"art")

art results are:
 visual art      7
artist          5
cover art       3
art             2
cover artist    1
cartographer    1
Name: 3rd_au_role, dtype: int64


In [561]:
change_formats(df,'1st_au_role',"lust|art|cover|photo","Illustrator")
change_formats(df,'2nd_au_role',"lust|art|cover|photo","Illustrator")
change_formats(df,'3rd_au_role',"lust|art|cover|photo","Illustrator")

Illustrator    19
Name: 1st_au_role, dtype: int64


/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


Illustrator    1461
Name: 2nd_au_role, dtype: int64
Illustrator    287
Name: 3rd_au_role, dtype: int64


In [562]:
find_formats(df,'1st_au_role',"lias")

lias results are:
 alias    10
Name: 1st_au_role, dtype: int64


In [563]:
change_formats(df,'1st_au_role',"name|lias","Pseudonym")
change_formats(df,'2nd_au_role',"name|lias","Pseudonym")
change_formats(df,'3rd_au_role',"name|lias","Pseudonym")

Series([], Name: 1st_au_role, dtype: int64)
Series([], Name: 2nd_au_role, dtype: int64)
Series([], Name: 3rd_au_role, dtype: int64)


In [564]:
find_formats(df,'1st_au_role',"word|face|ntro")

word|face|ntro results are:
 preface                  14
introduction             12
foreword                  6
selector/introduction     1
Name: 1st_au_role, dtype: int64


In [565]:
change_formats(df,'1st_au_role',"word|face|ntro|ward","Word")
change_formats(df,'2nd_au_role',"word|face|ntro|ward","Word")
change_formats(df,'3rd_au_role',"word|face|ntro|ward","Word")

Series([], Name: 1st_au_role, dtype: int64)
Series([], Name: 2nd_au_role, dtype: int64)
Series([], Name: 3rd_au_role, dtype: int64)


In [566]:
find_formats(df,'3rd_au_role',"ead")

ead results are:
 reading    218
read by     39
reader      34
Name: 3rd_au_role, dtype: int64


In [567]:
change_formats(df,'1st_au_role',"ead|erf|old|nar","Narrator")
change_formats(df,'2nd_au_role',"ead|erf|old|nar","Narrator")
change_formats(df,'3rd_au_role',"ead|erf|old|nar","Narrator")

Series([], Name: 1st_au_role, dtype: int64)
Series([], Name: 2nd_au_role, dtype: int64)
Series([], Name: 3rd_au_role, dtype: int64)


In [568]:
df['1st_au_role'].value_counts()[df['1st_au_role'].value_counts()>5]

               180478
Editor           1079
Pseudonym         537
Author            131
Adapter           106
Creator            68
Narrator           66
Contributor        53
Word               33
Illustrator        19
writer             12
Translator          8
writing as          6
compilation         6
Name: 1st_au_role, dtype: int64

In [569]:
find_formats(df,'1st_au_role',"with")

with results are:
 with    4
Name: 1st_au_role, dtype: int64


In [570]:
change_formats(df,'1st_au_role',"comp|with|prod","Contributor")
change_formats(df,'2nd_au_role',"comp|with|prod","Contributor")
change_formats(df,'3rd_au_role',"comp|with|prod","Contributor")

Series([], Name: 1st_au_role, dtype: int64)


/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


Series([], Name: 2nd_au_role, dtype: int64)
Series([], Name: 3rd_au_role, dtype: int64)


In [571]:
find_formats(df,'1st_au_role',"writ")

writ results are:
 writer                     12
writing as                  6
suspense writer             3
draft writer                1
written as mary roberts     1
Name: 1st_au_role, dtype: int64


In [572]:
change_formats(df,'1st_au_role',"writ|stor|atributed|as|novel|screenplay","Author")
change_formats(df,'2nd_au_role',"writ|stor|screenplay","Author")
change_formats(df,'3rd_au_role',"writ|stor|screenplay","Author")

Series([], Name: 1st_au_role, dtype: int64)
Series([], Name: 2nd_au_role, dtype: int64)
Series([], Name: 3rd_au_role, dtype: int64)


In [573]:
df['1st_au_role'].value_counts()[df['1st_au_role'].value_counts()>1]

                180478
Editor            1079
Pseudonym          537
Author             173
Adapter            106
Contributor         71
Creator             68
Narrator            66
Word                33
Illustrator         19
Translator           8
selections           3
abridged by          2
manufacturer         2
Name: 1st_au_role, dtype: int64

In [576]:
find_formats(df,'2nd_au_role',"by")

by results are:
 designed by             6
manufactured by         1
selected by             1
based on work by        1
prologue by             1
noted by                1
epilogue by             1
based on teleplay by    1
Name: 2nd_au_role, dtype: int64


Now, we set blanks to Author for au1 for all, and narrator for au2 for audio books

In [580]:
df['1st_au_role'][df['1st_au_role'] == ""] = "Author"

<ipython-input-580-9e3c9d84c6f4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['1st_au_role'][df['1st_au_role'] == ""] = "Author"


In [581]:
df['1st_au_role'].value_counts()[df['1st_au_role'].value_counts()>1]

Author          180651
Editor            1079
Pseudonym          537
Adapter            106
Contributor         71
Creator             68
Narrator            66
Word                33
Illustrator         19
Translator           8
selections           3
abridged by          2
manufacturer         2
Name: 1st_au_role, dtype: int64

In [584]:
df['2nd_au_role'][(df['2nd_au_role'] == "") & ((df.format == 'Audible') | (df.format == 'Audiobook'))] = 'Narrator'

<ipython-input-584-f6edc04f33e2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['2nd_au_role'][(df['2nd_au_role'] == "") & ((df.format == 'Audible') | (df.format == 'Audiobook'))] = 'Narrator'


In [587]:
df['2nd_au_role'][(df['2nd_au_role'] == "") & ((df.format != 'Audible') & (df.format != 'Audiobook'))] = 'Contributor'

<ipython-input-587-a76562edfc5e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['2nd_au_role'][(df['2nd_au_role'] == "") & ((df.format != 'Audible') & (df.format != 'Audiobook'))] = 'Contributor'


In [588]:
df['3rd_au_role'][(df['3rd_au_role'] == "")] = 'Contributor'

<ipython-input-588-210bf8e15223>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['3rd_au_role'][(df['3rd_au_role'] == "")] = 'Contributor'


In [592]:
df.head(10)

text_reviews_count   series country_code language_code is_ebook  \
0                   15   169353           US           NaN    false   
1                   60  1052227           US           eng     true   
2                   23   953679           US           eng     true   
3                    8   408775           US           NaN    false   
5                    3   199039           US           eng    false   
6                    5      NaN           US         en-GB     true   
7                    8   326237           US           NaN    false   
8                    6      NaN           US           eng    false   
9                    2   199041           US           eng    false   
10                  46   510553           US           eng    false   

    average_rating     format                 publisher num_pages  \
0             3.93  Hardcover       Simon & Schuster UK       400   
1             4.33      Ebook                       NaN       318   
2             3.49      Ebook   Amazon Digital Services       NaN   
3             3.30  Hardcover               Ulverscroft       NaN   
5             3.96  Audiobook               HarperAudio       NaN   
6             3.80      Ebook                       NaN       NaN   
7             3.61  Paperback                     Orion       320   
8             4.14  Paperback                 280 Steps       228   
9             3.93  Audiobook               HarperAudio       NaN   
10            4.07  Audiobook  Grand Central Publishing        13   

   publication_month  ... ratings_count   work_id  \
0                  4  ...           186   6243149   
1                NaN  ...           269  54143148   
2                NaN  ...           192  49305010   
3                 12  ...            52   1903897   
5                  3  ...             5   2288775   
6                NaN  ...             8  44866515   
7                  8  ...            58   2831381   
8                 11  ...            18  42251489   
9                  3  ...             5   6490729   
10                11  ...           250  24152432   

                                                title  \
0           Crowner Royal (Crowner John Mystery, #13)   
1             The House of Memory (Pluto's Snitch #2)   
2   The Slaughtered Virgin of Zenopolis (Inspector...   
3             Dead in the Morning (Patrick Grant, #1)   
5                               A Murder is Announced   
6                                  Dark Flames Rising   
7                     Wycliffe and the Cycle of Death   
8                          The Cost of Doing Business   
9                                4:50 From Paddington   
10                       White Fire (Pendergast, #13)   

                                 title_without_series  1st_au_id 1st_au_role  \
0           Crowner Royal (Crowner John Mystery, #13)      37778      Author   
1             The House of Memory (Pluto's Snitch #2)     242185      Author   
2   The Slaughtered Virgin of Zenopolis (Inspector...   15104629      Author   
3             Dead in the Morning (Patrick Grant, #1)     190988      Author   
5                               A Murder is Announced     123715      Author   
6                                  Dark Flames Rising   13658169      Author   
7                     Wycliffe and the Cycle of Death    1533165      Author   
8                          The Cost of Doing Business    4577517      Author   
9                                4:50 From Paddington     123715      Author   
10                       White Fire (Pendergast, #13)      12577      Author   

   2nd_au_id  2nd_au_role 3rd_au_id 3rd_au_role  
0        NaN          NaN       NaN         NaN  
1        NaN          NaN       NaN         NaN  
2   14425713  Contributor       NaN         NaN  
3        NaN          NaN       NaN         NaN  
5     552417     Narrator       NaN         NaN  
6        NaN          NaN       NaN         NaN  
7        NaN       

## Main data wrangling complete. 

In [597]:
df.to_pickle('./Books_wrangled.pkl')

Load author avg rating, ratings cont, and review count to join later on. 

In [600]:
df_authors = pd.read_json('goodreads_book_authors.json')

In [601]:
df_authors.head()

average_rating  author_id  text_reviews_count              name  \
0            3.98     604031                   7  Ronald J. Fields   
1            4.08     626222               28716     Anita Diamant   
2            3.92      10333                5075    Barbara Hambly   
3            3.68       9212               36262   Jennifer Weiner   
4            3.82     149918                  96     Nigel Pennick   

   ratings_count  
0             49  
1         546796  
2         122118  
3         888522  
4           1740

In [603]:
df_authors.drop(columns=['name'], inplace=True)

In [604]:
df_authors.isnull().sum()

average_rating        0
author_id             0
text_reviews_count    0
ratings_count         0
dtype: int64

In [605]:
df_authors.to_pickle('./Authors_wrangled.pkl')

Loading series data

In [608]:
df_series = pd.read_json('goodreads_book_series.json')

In [609]:
df_series.head()

numbered note                                        description  \
0     true                                                           
1     true       This series is also known as * Avalon : Jalina...   
2     true       Plot-wise, "Crowner's Crusade" is a prequel to...   
3     true                                                           
4    false                                                           

                                      title  series_works_count  series_id  \
0                     Sun Wolf and Starhawk                   9     189911   
1                      Avalon: Web of Magic                  14     151854   
2                      Crowner John Mystery                  15     169353   
3                            Pluto's Snitch                   3    1052227   
4  The Way, With The Enchanted Neighborhood                   3    1070125   

   primary_work_count  
0                   3  
1                  12  
2                  15  
3                   3  
4                   3

In [611]:
df_series.drop(columns=['description', 'note','title'], inplace=True)

In [612]:
df_series.isnull().sum()

numbered              0
series_works_count    0
series_id             0
primary_work_count    0
dtype: int64

In [613]:
df_series.to_pickle('./Series_wrangled.pkl')

Loading works data (has best id in there if need to slim)

In [614]:
df_works = pd.read_json('goodreads_book_works.json')

In [615]:
df_works.head()

books_count  reviews_count original_publication_month  \
0            1              6                          8   
1           22          10162                              
2            2            268                              
3           38          89252                          7   
4            2             49                              

  default_description_language_code  text_reviews_count  best_book_id  \
0                                                     1       5333265   
1                                                   741         25717   
2                                                     7       7327624   
3                                                  3504       6066819   
4                                                     5        287140   

  original_publication_year  \
0                      1984   
1                      2001   
2                      1987   
3                      2009   
4                      1990   

                                      original_title  \
0                       W. C. Fields: A Life on Film   
1                                        Good Harbor   
2                                                      
3                               Best Friends Forever   
4  Runic Astrology: Starcraft and Timekeeping in ...   

                                        rating_dist  \
0                       5:1|4:1|3:1|2:0|1:0|total:3   
1        5:517|4:1787|3:2763|2:966|1:196|total:6229   
2                  5:49|4:58|3:26|2:5|1:3|total:141   
3  5:9152|4:16855|3:19507|2:6210|1:1549|total:53273   
4                      5:6|4:1|3:3|2:3|1:2|total:15   

  default_chaptering_book_id original_publication_day original_language_id  \
0                                                                            
1                                                                            
2                                                                            
3                                                  14                        
4                                                                            

   ratings_count media_type  ratings_sum  work_id  
0              3       book           12  5400751  
1           6229       book        20150  1323437  
2            141       book          568  8948723  
3          53273       book       185670  6243154  
4             15       book           51   278577

In [617]:
df_works.drop(columns=['default_description_language_code','default_chaptering_book_id','original_publication_day','original_language_id'],inplace=True)

In [618]:
#This is deceiving...there are blanks
df_works.isnull().sum()

books_count                   0
reviews_count                 0
original_publication_month    0
text_reviews_count            0
best_book_id                  0
original_publication_year     0
original_title                0
rating_dist                   0
ratings_count                 0
media_type                    0
ratings_sum                   0
work_id                       0
dtype: int64

In [619]:
df_works.to_pickle('./Works_wrangled.pkl')

Looking at reviewer data

In [4]:
df_revs = pd.read_json('goodreads_reviews_mystery_thriller_crime.json')

In [5]:
df_revs.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d   6392944   
1  8842281e1d1347389f2ab93d60773d4d  28684704   
2  8842281e1d1347389f2ab93d60773d4d  32283133   
3  8842281e1d1347389f2ab93d60773d4d  17860739   
4  8842281e1d1347389f2ab93d60773d4d   8694005   

                          review_id  rating  \
0  5e212a62bced17b4dbe41150e5bb9037       3   
1  2ede853b14dc4583f96cf5d120af636f       3   
2  8e4d61801907e591018bdc3442a9cf2b       0   
3  022bb6daffa49adc27f6b20b6ebeb37d       4   
4  0e317947e1fd341f573192111bb2921d       3   

                                         review_text  \
0  I haven't read a fun mystery book in a while a...   
1  A fun, fast paced science fiction thriller. I ...   
2           http://www.telegraph.co.uk/culture/10...   
3  An amazing and unique creation: JJ Abrams and ...   
4  The Name of the Rose is a thrilling Dan Brown-...   

                       date_added                    date_updated  \
0  Mon Jul 24 02:48:17 -0700 2017  Sun Jul 30 09:28:03 -0700 2017   
1  Tue Nov 15 11:29:22 -0800 2016  Mon Mar 20 23:40:27 -0700 2017   
2  Tue Nov 01 11:09:18 -0700 2016  Tue Nov 01 11:09:44 -0700 2016   
3  Wed Mar 26 13:51:30 -0700 2014  Tue Sep 23 01:44:36 -0700 2014   
4  Wed Sep 08 01:22:27 -0700 2010  Wed Dec 14 12:30:43 -0800 2016   

                          read_at                      started_at  n_votes  \
0  Tue Jul 25 00:00:00 -0700 2017  Mon Jul 24 00:00:00 -0700 2017        6   
1  Sat Mar 18 23:22:42 -0700 2017  Fri Mar 17 23:45:40 -0700 2017       22   
2                                                                        9   
3  Sun Sep 21 00:00:00 -0700 2014  Sat Jul 26 00:00:00 -0700 2014        7   
4  Mon Aug 10 00:00:00 -0700 2015  Mon Jul 20 00:00:00 -0700 2015       17   

   n_comments  
0           0  
1           0  
2           0  
3           0  
4           6

In [6]:
df_revs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1849236 entries, 0 to 1849235
Data columns (total 11 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   user_id       object
 1   book_id       int64 
 2   review_id     object
 3   rating        int64 
 4   review_text   object
 5   date_added    object
 6   date_updated  object
 7   read_at       object
 8   started_at    object
 9   n_votes       int64 
 10  n_comments    int64 
dtypes: int64(4), object(7)
memory usage: 155.2+ MB


In [7]:
#df_revs['review_word_count'] = df_revs['review_text'].str.split().str.len()

In [9]:
df= pd.read_pickle('./Books_wrangled.pkl')

In [16]:
df.book_id.nunique()

182663

In [10]:
lt = df.book_id.unique()

In [11]:
lt

array([ 6066814, 33394837, 29074697, ..., 25727303, 23252156, 26168430])

In [12]:
df_revs_slim = df_revs[df_revs['book_id'].isin(lt)]

In [13]:
df_revs_slim.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1742885 entries, 0 to 1849235
Data columns (total 11 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   user_id       object
 1   book_id       int64 
 2   review_id     object
 3   rating        int64 
 4   review_text   object
 5   date_added    object
 6   date_updated  object
 7   read_at       object
 8   started_at    object
 9   n_votes       int64 
 10  n_comments    int64 
dtypes: int64(4), object(7)
memory usage: 159.6+ MB


In [15]:
df_revs_slim.book_id.nunique()

182491

In [ ]:
#df_revs_slim['review_word_count'] = df_revs_slim['review_text'].str.split().str.len()

In [33]:
n=20000
list_df_rs = [df_revs_slim[i:i+n] for i in range(0,df_revs_slim.shape[0],n)]

In [34]:
len(list_df_rs)

88

In [38]:
for i in range(88):
    list_df_rs[i]['wc'] = list_df_rs[i]['review_text'].str.split().str.len()

<ipython-input-38-416ca43bfea7>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  list_df_rs[i]['wc'] = list_df_rs[i]['review_text'].str.split().str.len()


In [41]:
list_df_rs[87]

user_id   book_id  \
1846170  b19d27dfed190cfa2ccc7de96cb11423    148144   
1846171  b19d27dfed190cfa2ccc7de96cb11423    136636   
1846172  b19d27dfed190cfa2ccc7de96cb11423     39033   
1846173  b19d27dfed190cfa2ccc7de96cb11423    993736   
1846174  b19d27dfed190cfa2ccc7de96cb11423       968   
...                                   ...       ...   
1849231  c5ae39ab55f5031b440d93f3f09ee736  13263377   
1849232  c5ae39ab55f5031b440d93f3f09ee736  15999321   
1849233  c5ae39ab55f5031b440d93f3f09ee736  13261268   
1849234  c5ae39ab55f5031b440d93f3f09ee736  15948118   
1849235  8ba77e3c745ebddccc6306fc3c6bb25e  17411706   

                                review_id  rating  \
1846170  202e6dae93ba3abda3e0a38d770343de       4   
1846171  c2ccb4990f0dc772679202c76f896522       3   
1846172  d9ab2cf56470b207c606e083b40b281c       4   
1846173  5bcace415277a92a1f7190d1cd1d8b3b       5   
1846174  02629d67d65200638b7d662a4e658dab       0   
...                                   ...     ...   
1849231  97569c40e28cac15ad5c69a5355d4da3       5   
1849232  3827b3889af45bb11f0d8bc1b71b4f3e       5   
1849233  494baf5b0cfd4f8bf7d6f405ec6895f8       5   
1849234  4c0ab59da1339bde2a94cf36148a46d9       0   
1849235  5a6c09f228988f0b69e23f50f39575ed       4   

                                               review_text  \
1846170  stephanie Plum is the greatest real crazy crea...   
1846171  This is a archaeological expedition with hidde...   
1846172  The FBI agent in the story is better than Jame...   
1846173  This is my favorite book of all times. I would...   
1846174  The way Dan Brown wrote this story was excitin...   
...                                                    ...   
1849231  Double Exposure by Susan Sleeman is a well-tho...   
1849232  Are you a suspense fan? Do you crave action an...   
1849233  Broken Trust by Sharon Dunn is a romantic susp...   
1849234  Scrambled was a good title for the book as scr...   
1849235  I'm not a regular mystery reader, however, thi...   

                             date_added                    date_updated  \
1846170  Tue Feb 19 16:40:55 -0800 2008  Tue Feb 19 16:44:53 -0800 2008   
1846171  Tue Feb 19 16:19:33 -0800 2008  Tue Feb 19 16:27:12 -0800 2008   
1846172  Sat Feb 09 05:21:53 -0800 2008  Sat Feb 09 05:28:53 -0800 2008   
1846173  Sat Feb 09 05:16:35 -0800 2008  Mon Jul 05 08:50:17 -0700 2010   
1846174  Sat Feb 09 05:11:23 -0800 2008  Sat Feb 09 05:16:02 -0800 2008   
...                                 ...                             ...   
1849231  Fri Jun 28 11:44:01 -0700 2013  Fri Jun 28 11:44:41 -0700 2013   
1849232  Fri Jun 28 11:42:43 -0700 2013  Fri Jun 28 11:42:53 -0700 2013   
1849233  Fri Jun 28 09:07:53 -0700 2013  Fri Jun 28 09:08:21 -0700 2013   
1849234  Fri Sep 14 19:57:37 -0700 2012  Sat Jan 17 08:54:31 -0800 2015   
1849235  Wed Jul 03 05:09:04 -0700 2013   Wed Jul 03 05:11:43 -07002013   

                                read_at started_at  n_votes  n_comments   wc  
1846170  Thu Sep 09 00:00:00 -0700 2004                   0           1   78  
1846171  Sun Feb 17 00:00:00 -0800 2008                   0           0   58  
1846172  Mon Jan 07 00:00:00 -0800 2008                   0           0   71  
1846173  Sun Sep 16 00:00:00 -0700 2001                   0           0   53  
1846174  Wed Aug 10 00:00:00 -0700 2005                   0           0   53  
...                                 ...        ...      ...         ...  ...  
1849231                                                   0           0  130  
1849232                                                   0           0  109  
1849233                                                   0           0  285  
1849234                                                   0           0  174  
1849235                                                   0           0  238  

[2885 rows x 12 columns]

In [48]:
new_df = pd.DataFrame()

for i in range(88):
    new_df = pd.concat([new_df,list_df_rs[i]])

In [49]:
new_df

user_id   book_id  \
0        8842281e1d1347389f2ab93d60773d4d   6392944   
1        8842281e1d1347389f2ab93d60773d4d  28684704   
2        8842281e1d1347389f2ab93d60773d4d  32283133   
3        8842281e1d1347389f2ab93d60773d4d  17860739   
4        8842281e1d1347389f2ab93d60773d4d   8694005   
...                                   ...       ...   
1849231  c5ae39ab55f5031b440d93f3f09ee736  13263377   
1849232  c5ae39ab55f5031b440d93f3f09ee736  15999321   
1849233  c5ae39ab55f5031b440d93f3f09ee736  13261268   
1849234  c5ae39ab55f5031b440d93f3f09ee736  15948118   
1849235  8ba77e3c745ebddccc6306fc3c6bb25e  17411706   

                                review_id  rating  \
0        5e212a62bced17b4dbe41150e5bb9037       3   
1        2ede853b14dc4583f96cf5d120af636f       3   
2        8e4d61801907e591018bdc3442a9cf2b       0   
3        022bb6daffa49adc27f6b20b6ebeb37d       4   
4        0e317947e1fd341f573192111bb2921d       3   
...                                   ...     ...   
1849231  97569c40e28cac15ad5c69a5355d4da3       5   
1849232  3827b3889af45bb11f0d8bc1b71b4f3e       5   
1849233  494baf5b0cfd4f8bf7d6f405ec6895f8       5   
1849234  4c0ab59da1339bde2a94cf36148a46d9       0   
1849235  5a6c09f228988f0b69e23f50f39575ed       4   

                                               review_text  \
0        I haven't read a fun mystery book in a while a...   
1        A fun, fast paced science fiction thriller. I ...   
2                 http://www.telegraph.co.uk/culture/10...   
3        An amazing and unique creation: JJ Abrams and ...   
4        The Name of the Rose is a thrilling Dan Brown-...   
...                                                    ...   
1849231  Double Exposure by Susan Sleeman is a well-tho...   
1849232  Are you a suspense fan? Do you crave action an...   
1849233  Broken Trust by Sharon Dunn is a romantic susp...   
1849234  Scrambled was a good title for the book as scr...   
1849235  I'm not a regular mystery reader, however, thi...   

                             date_added                    date_updated  \
0        Mon Jul 24 02:48:17 -0700 2017  Sun Jul 30 09:28:03 -0700 2017   
1        Tue Nov 15 11:29:22 -0800 2016  Mon Mar 20 23:40:27 -0700 2017   
2        Tue Nov 01 11:09:18 -0700 2016  Tue Nov 01 11:09:44 -0700 2016   
3        Wed Mar 26 13:51:30 -0700 2014  Tue Sep 23 01:44:36 -0700 2014   
4        Wed Sep 08 01:22:27 -0700 2010  Wed Dec 14 12:30:43 -0800 2016   
...                                 ...                             ...   
1849231  Fri Jun 28 11:44:01 -0700 2013  Fri Jun 28 11:44:41 -0700 2013   
1849232  Fri Jun 28 11:42:43 -0700 2013  Fri Jun 28 11:42:53 -0700 2013   
1849233  Fri Jun 28 09:07:53 -0700 2013  Fri Jun 28 09:08:21 -0700 2013   
1849234  Fri Sep 14 19:57:37 -0700 2012  Sat Jan 17 08:54:31 -0800 2015   
1849235  Wed Jul 03 05:09:04 -0700 2013   Wed Jul 03 05:11:43 -07002013   

                                read_at                      started_at  \
0        Tue Jul 25 00:00:00 -0700 2017  Mon Jul 24 00:00:00 -0700 2017   
1        Sat Mar 18 23:22:42 -0700 2017  Fri Mar 17 23:45:40 -0700 2017   
2                                                                         
3        Sun Sep 21 00:00:00 -0700 2014  Sat Jul 26 00:00:00 -0700 2014   
4        Mon Aug 10 00:00:00 -0700 2015  Mon Jul 20 00:00:00 -0700 2015   
...                                 ...                             ...   
1849231                                                                   
1849232                                                                   
1849233                                                                   
1849234                                                                   
1849235                                                                   

         n_votes  n_comments   wc  
0              6           0   86  
1             22           0  465  
2              9           0    1  
3              7           0  483  
4        

In [50]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1742885 entries, 0 to 1849235
Data columns (total 12 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   user_id       object
 1   book_id       int64 
 2   review_id     object
 3   rating        int64 
 4   review_text   object
 5   date_added    object
 6   date_updated  object
 7   read_at       object
 8   started_at    object
 9   n_votes       int64 
 10  n_comments    int64 
 11  wc            int64 
dtypes: int64(5), object(7)
memory usage: 172.9+ MB


In [51]:
new_df.drop(columns=['review_text'], inplace=True)

In [52]:
#new_df.to_pickle('Reviews_mostlyWrangled.pkl')

In [63]:
pd.to_datetime(list_df_rs[80].date_added, infer_datetime_format=True, errors='coerce')

1694005    2013-08-18 08:13:04-07:00
1694006    2013-06-23 15:39:18-07:00
1694007    2012-06-29 05:55:06-07:00
1694008    2017-08-25 19:28:37-07:00
1694009    2017-08-01 17:07:51-07:00
                     ...            
1714635    2014-07-19 12:30:37-07:00
1714636    2014-06-29 07:10:52-07:00
1714637    2014-06-25 07:11:54-07:00
1714638    2014-04-07 18:04:16-07:00
1714639    2014-03-07 10:09:43-08:00
Name: date_added, Length: 20000, dtype: object

In [64]:
for i in range(88):
    list_df_rs[i].date_added = pd.to_datetime(list_df_rs[i].date_added, infer_datetime_format=True, errors='coerce')

/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py:5159: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [71]:
list_df_rs[10].head()

user_id   book_id  \
210150  faf123d9fef513663c8257f60f5f7abb    446038   
210151  faf123d9fef513663c8257f60f5f7abb     32428   
210152  326c68bca9725ce3da07d3bc9b12dfa3  16290044   
210153  326c68bca9725ce3da07d3bc9b12dfa3  16123327   
210154  326c68bca9725ce3da07d3bc9b12dfa3  17791326   

                               review_id  rating  \
210150  9ffa02c0f5a643f6aa225c69bd87a318       3   
210151  1310349035f90897602033cc50b612f0       3   
210152  af0d518d4ff323c5ee7f59c39986c5ea       1   
210153  45906c46f95dd1f582e37701eba74cf8       4   
210154  0c94593d51172f8778914913a1339faf       4   

                                              review_text  \
210150  I expected more of an Edgar Award Winner. The ...   
210151  Meh. In my opinion you can't go wrong with Koo...   
210152  The premise of this book really interested me....   
210153  Many people like getting scared by horror, bec...   
210154  This book kept me up until 4 in the morning be...   

                       date_added               date_updated  \
210150  2008-07-30 20:16:14-07:00  2011-03-11 08:27:26-08:00   
210151  2008-07-30 20:15:13-07:00  2010-03-30 09:20:50-07:00   
210152  2014-03-21 04:49:22-07:00  2014-06-29 10:02:24-07:00   
210153  2013-05-20 07:58:25-07:00  2013-07-20 13:44:37-07:00   
210154  2013-05-19 07:56:39-07:00  2013-06-22 10:48:48-07:00   

                          read_at                 started_at  n_votes  \
210150  2010-03-29 00:00:00-07:00                        NaT        0   
210151  2010-03-30 00:00:00-07:00                        NaT        0   
210152                        NaT                        NaT        1   
210153  2013-07-20 13:44:37-07:00  2013-07-09 00:00:00-07:00        2   
210154  2013-06-22 10:48:48-07:00  2013-06-03 00:00:00-07:00        0   

        n_comments   wc  
210150           0   47  
210151           0   16  
210152           0  242  
210153           0  218  
210154           0  251

In [66]:
for i in range(88):
    list_df_rs[i].date_updated = pd.to_datetime(list_df_rs[i].date_updated, infer_datetime_format=True, errors='coerce')

/Users/jensensj/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py:5159: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [67]:
for i in range(88):
    list_df_rs[i].read_at = pd.to_datetime(list_df_rs[i].read_at, infer_datetime_format=True, errors='coerce')

In [68]:
for i in range(88):
    list_df_rs[i].started_at = pd.to_datetime(list_df_rs[i].started_at, infer_datetime_format=True, errors='coerce')

In [72]:
new_df_2 = pd.DataFrame()

for i in range(88):
    new_df_2 = pd.concat([new_df_2,list_df_rs[i]])

In [73]:
new_df_2.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d   6392944   
1  8842281e1d1347389f2ab93d60773d4d  28684704   
2  8842281e1d1347389f2ab93d60773d4d  32283133   
3  8842281e1d1347389f2ab93d60773d4d  17860739   
4  8842281e1d1347389f2ab93d60773d4d   8694005   

                          review_id  rating  \
0  5e212a62bced17b4dbe41150e5bb9037       3   
1  2ede853b14dc4583f96cf5d120af636f       3   
2  8e4d61801907e591018bdc3442a9cf2b       0   
3  022bb6daffa49adc27f6b20b6ebeb37d       4   
4  0e317947e1fd341f573192111bb2921d       3   

                                         review_text  \
0  I haven't read a fun mystery book in a while a...   
1  A fun, fast paced science fiction thriller. I ...   
2           http://www.telegraph.co.uk/culture/10...   
3  An amazing and unique creation: JJ Abrams and ...   
4  The Name of the Rose is a thrilling Dan Brown-...   

                  date_added               date_updated  \
0  2017-07-24 02:48:17-07:00  2017-07-30 09:28:03-07:00   
1  2016-11-15 11:29:22-08:00  2017-03-20 23:40:27-07:00   
2  2016-11-01 11:09:18-07:00  2016-11-01 11:09:44-07:00   
3  2014-03-26 13:51:30-07:00  2014-09-23 01:44:36-07:00   
4  2010-09-08 01:22:27-07:00  2016-12-14 12:30:43-08:00   

                     read_at                 started_at  n_votes  n_comments  \
0  2017-07-25 00:00:00-07:00  2017-07-24 00:00:00-07:00        6           0   
1  2017-03-18 23:22:42-07:00  2017-03-17 23:45:40-07:00       22           0   
2                        NaT                        NaT        9           0   
3  2014-09-21 00:00:00-07:00  2014-07-26 00:00:00-07:00        7           0   
4  2015-08-10 00:00:00-07:00  2015-07-20 00:00:00-07:00       17           6   

    wc  
0   86  
1  465  
2    1  
3  483  
4  376

In [77]:
new_df_2.drop(columns=['review_text'], inplace=True)

In [78]:
new_df_2.to_pickle('Reviews_Wrangled.pkl')

Now on to interactions

c_size = 10**4
df_ints_iter = pd.read_json('goodreads_interactions_mystery_thriller_crime.json', orient='records',lines=True,chunksize=c_size)

print(df_ints_iter)



for i in df_ints_iter:
    print(i)

In [ ]:
df_ints = pd.read_json('goodreads_interactions_mystery_thriller_crime.json', lines=True)

In [ ]:
#df_iter = pd.read_json('inter_test.json', lines=True,chunksize=10)


In [2]:
with open('goodreads_interactions_mystery_thriller_crime.json') as json_file:      
    data = json_file.readlines()
    # this line below may take at least 8-10 minutes of processing for 4-5 million rows. It converts all strings in list to actual json objects. 
    data = list(map(json.loads, data)) 

In [3]:
#data

In [4]:
data_t = data

In [5]:
#del data_t['review_text_incomplete']

In [12]:
data_t[4]['review_text_incomplete']

'A fun, fast paced science fiction thriller. I read it in 2 nights and couldn\'t put it down. The book is about the quantum theory of <a href="https://en.wikipedia.org/wiki/Many-worlds_interpretation" rel="nofollow">many worlds</a> which states that all decisions we make throughout our lives basically create branches, and that each possible path through the decision tree can be thou...'

In [14]:
df_ints = pd.DataFrame(data)

In [15]:
df_ints.columns


Index(['user_id', 'book_id', 'review_id', 'is_read', 'rating',
       'review_text_incomplete', 'date_added', 'date_updated', 'read_at',
       'started_at'],
      dtype='object')

In [17]:
df_ints.drop(columns=['review_text_incomplete'],inplace=True)

In [2]:
df_ints.columns

NameError: name 'df_ints' is not defined

In [ ]:
df_ints.to_pickle('Ints_SortOfWrangled.pkl')

In [3]:
ints_0 = pd.read_json('./splits_x/xaa.json',lines=True)

In [5]:
 ints_0.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d   6392944   
1  8842281e1d1347389f2ab93d60773d4d   2279538   
2  8842281e1d1347389f2ab93d60773d4d  20821043   
3  8842281e1d1347389f2ab93d60773d4d  31184479   
4  8842281e1d1347389f2ab93d60773d4d  28684704   

                          review_id  is_read  rating  \
0  5e212a62bced17b4dbe41150e5bb9037     True       3   
1  556e9e3f00fb5b7e4eaa116764b771b0    False       0   
2  4b15f8a08f04a17f9a47f2b76fec21e5    False       0   
3  0b9fc382ad38c5804ec46df84d66be6e    False       0   
4  2ede853b14dc4583f96cf5d120af636f     True       3   

                              review_text_incomplete  \
0  I haven't read a fun mystery book in a while a...   
1                                                      
2                                                      
3                                                      
4  A fun, fast paced science fiction thriller. I ...   

                       date_added                    date_updated  \
0  Mon Jul 24 02:48:17 -0700 2017  Sun Jul 30 09:28:03 -0700 2017   
1  Wed Mar 29 00:27:14 -0700 2017  Wed Mar 29 00:27:14 -0700 2017   
2  Mon Mar 27 22:52:11 -0700 2017  Mon Mar 27 22:52:12 -0700 2017   
3  Fri Jan 06 16:22:46 -0800 2017  Fri Jan 06 16:22:46 -0800 2017   
4  Tue Nov 15 11:29:22 -0800 2016  Mon Mar 20 23:40:27 -0700 2017   

                          read_at                      started_at  
0  Tue Jul 25 00:00:00 -0700 2017  Mon Jul 24 00:00:00 -0700 2017  
1                                                                  
2                                                                  
3                                                                  
4  Sat Mar 18 23:22:42 -0700 2017  Fri Mar 17 23:45:40 -0700 2017

In [7]:
ints_0.drop(columns=['review_text_incomplete'],inplace=True)

In [8]:
ints_0.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d   6392944   
1  8842281e1d1347389f2ab93d60773d4d   2279538   
2  8842281e1d1347389f2ab93d60773d4d  20821043   
3  8842281e1d1347389f2ab93d60773d4d  31184479   
4  8842281e1d1347389f2ab93d60773d4d  28684704   

                          review_id  is_read  rating  \
0  5e212a62bced17b4dbe41150e5bb9037     True       3   
1  556e9e3f00fb5b7e4eaa116764b771b0    False       0   
2  4b15f8a08f04a17f9a47f2b76fec21e5    False       0   
3  0b9fc382ad38c5804ec46df84d66be6e    False       0   
4  2ede853b14dc4583f96cf5d120af636f     True       3   

                       date_added                    date_updated  \
0  Mon Jul 24 02:48:17 -0700 2017  Sun Jul 30 09:28:03 -0700 2017   
1  Wed Mar 29 00:27:14 -0700 2017  Wed Mar 29 00:27:14 -0700 2017   
2  Mon Mar 27 22:52:11 -0700 2017  Mon Mar 27 22:52:12 -0700 2017   
3  Fri Jan 06 16:22:46 -0800 2017  Fri Jan 06 16:22:46 -0800 2017   
4  Tue Nov 15 11:29:22 -0800 2016  Mon Mar 20 23:40:27 -0700 2017   

                          read_at                      started_at  
0  Tue Jul 25 00:00:00 -0700 2017  Mon Jul 24 00:00:00 -0700 2017  
1                                                                  
2                                                                  
3                                                                  
4  Sat Mar 18 23:22:42 -0700 2017  Fri Mar 17 23:45:40 -0700 2017

In [9]:
ints_0.date_added = pd.to_datetime(ints_0.date_added,infer_datetime_format=True, errors='coerce')

In [10]:
ints_0.date_added

0         2017-07-24 02:48:17-07:00
1         2017-03-29 00:27:14-07:00
2         2017-03-27 22:52:11-07:00
3         2017-01-06 16:22:46-08:00
4         2016-11-15 11:29:22-08:00
                    ...            
999995    2013-09-13 15:48:48-07:00
999996    2013-04-13 17:07:56-07:00
999997    2013-03-08 04:22:15-08:00
999998    2013-03-08 04:22:06-08:00
999999    2013-02-24 13:51:33-08:00
Name: date_added, Length: 1000000, dtype: object

Run command line: split -l 1000000 goodreads_interactions_mystery_thriller_crime.json

In [32]:
df_i_0 = pd.DataFrame()
for i in range(5):
    filename = './splits_x/'+str(i)+'.json'
    df_s = pd.read_json(filename,lines=True)
    df_s.drop(columns=['review_text_incomplete'],inplace=True)
    df_s.date_added = pd.to_datetime(df_s.date_added,infer_datetime_format=True, errors='coerce')
    df_s.date_updated = pd.to_datetime(df_s.date_updated,infer_datetime_format=True, errors='coerce')
    df_s.read_at = pd.to_datetime(df_s.read_at,infer_datetime_format=True, errors='coerce')
    df_s.started_at = pd.to_datetime(df_s.started_at,infer_datetime_format=True, errors='coerce')
    df_i_0 = pd.concat([df_i_0,df_s])

In [33]:
df_i_0.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d   6392944   
1  8842281e1d1347389f2ab93d60773d4d   2279538   
2  8842281e1d1347389f2ab93d60773d4d  20821043   
3  8842281e1d1347389f2ab93d60773d4d  31184479   
4  8842281e1d1347389f2ab93d60773d4d  28684704   

                          review_id  is_read  rating  \
0  5e212a62bced17b4dbe41150e5bb9037     True       3   
1  556e9e3f00fb5b7e4eaa116764b771b0    False       0   
2  4b15f8a08f04a17f9a47f2b76fec21e5    False       0   
3  0b9fc382ad38c5804ec46df84d66be6e    False       0   
4  2ede853b14dc4583f96cf5d120af636f     True       3   

                  date_added               date_updated  \
0  2017-07-24 02:48:17-07:00  2017-07-30 09:28:03-07:00   
1  2017-03-29 00:27:14-07:00  2017-03-29 00:27:14-07:00   
2  2017-03-27 22:52:11-07:00  2017-03-27 22:52:12-07:00   
3  2017-01-06 16:22:46-08:00  2017-01-06 16:22:46-08:00   
4  2016-11-15 11:29:22-08:00  2017-03-20 23:40:27-07:00   

                     read_at                 started_at  
0  2017-07-25 00:00:00-07:00  2017-07-24 00:00:00-07:00  
1                        NaT                        NaT  
2                        NaT                        NaT  
3                        NaT                        NaT  
4  2017-03-18 23:22:42-07:00  2017-03-17 23:45:40-07:00

In [34]:
df_i_0.to_pickle('df_i_0.pkl')

In [36]:
df_i_1 = pd.DataFrame()
for i in range(5,10):
    filename = './splits_x/'+str(i)+'.json'
    df_s = pd.read_json(filename,lines=True)
    df_s.drop(columns=['review_text_incomplete'],inplace=True)
    df_s.date_added = pd.to_datetime(df_s.date_added,infer_datetime_format=True, errors='coerce')
    df_s.date_updated = pd.to_datetime(df_s.date_updated,infer_datetime_format=True, errors='coerce')
    df_s.read_at = pd.to_datetime(df_s.read_at,infer_datetime_format=True, errors='coerce')
    df_s.started_at = pd.to_datetime(df_s.started_at,infer_datetime_format=True, errors='coerce')
    df_i_1 = pd.concat([df_i_1,df_s])

In [37]:
df_i_1.to_pickle('df_i_1.pkl')

In [39]:
df_i_2 = pd.DataFrame()
for i in range(10,15):
    filename = './splits_x/'+str(i)+'.json'
    df_s = pd.read_json(filename,lines=True)
    df_s.drop(columns=['review_text_incomplete'],inplace=True)
    df_s.date_added = pd.to_datetime(df_s.date_added,infer_datetime_format=True, errors='coerce')
    df_s.date_updated = pd.to_datetime(df_s.date_updated,infer_datetime_format=True, errors='coerce')
    df_s.read_at = pd.to_datetime(df_s.read_at,infer_datetime_format=True, errors='coerce')
    df_s.started_at = pd.to_datetime(df_s.started_at,infer_datetime_format=True, errors='coerce')
    df_i_2 = pd.concat([df_i_2,df_s])

In [40]:
df_i_2.to_pickle('df_i_2.pkl')

In [41]:
df_i_3 = pd.DataFrame()
for i in range(15,20):
    filename = './splits_x/'+str(i)+'.json'
    df_s = pd.read_json(filename,lines=True)
    df_s.drop(columns=['review_text_incomplete'],inplace=True)
    df_s.date_added = pd.to_datetime(df_s.date_added,infer_datetime_format=True, errors='coerce')
    df_s.date_updated = pd.to_datetime(df_s.date_updated,infer_datetime_format=True, errors='coerce')
    df_s.read_at = pd.to_datetime(df_s.read_at,infer_datetime_format=True, errors='coerce')
    df_s.started_at = pd.to_datetime(df_s.started_at,infer_datetime_format=True, errors='coerce')
    df_i_3 = pd.concat([df_i_3,df_s])

In [42]:
df_i_3.to_pickle('df_i_3.pkl')

In [43]:
df_i_4 = pd.DataFrame()
for i in range(20,25):
    filename = './splits_x/'+str(i)+'.json'
    df_s = pd.read_json(filename,lines=True)
    df_s.drop(columns=['review_text_incomplete'],inplace=True)
    df_s.date_added = pd.to_datetime(df_s.date_added,infer_datetime_format=True, errors='coerce')
    df_s.date_updated = pd.to_datetime(df_s.date_updated,infer_datetime_format=True, errors='coerce')
    df_s.read_at = pd.to_datetime(df_s.read_at,infer_datetime_format=True, errors='coerce')
    df_s.started_at = pd.to_datetime(df_s.started_at,infer_datetime_format=True, errors='coerce')
    df_i_4 = pd.concat([df_i_4,df_s])

In [44]:
df_i_4.to_pickle('df_i_4.pkl')